File: concepts/06_text_binary_and_serialization/05_serialization_clone_and_transfer/nodejs/test_02_numeric_and_trust_boundaries.mjs
// 数值模型、类型恢复与信任边界。
// 共同问题：通用数据格式是否保留语言数值类型；非标准数值如何处理；
// 自定义反序列化如何限制可构造类型和字段。
//
// polyglot-family: text_binary_and_serialization
// polyglot-concept: serialization_clone_and_transfer
// polyglot-related: languages/nodejs/language/test_020_json_serialization_parsing_and_structured_clone.mjs

import assert from 'node:assert/strict';
import test from 'node:test';

test('JSON number 超过安全整数范围会按 Number 模型丢失精度', () => {
  const text = '9007199254740993';
  const decoded = JSON.parse(text);

  assert.equal(decoded, 9007199254740992);
  assert.notEqual(BigInt(decoded), BigInt(text));
  assert.throws(() => JSON.stringify(9007199254740993n), TypeError);

  // 若协议需要该整数，应编码为带规则的字符串并显式恢复 BigInt，不能静默 round trip。
});

test('reviver 只恢复白名单标签和字段', () => {
  function restore(key, value) {
    if (value?.type !== 'Point') {
      return value;
    }
    if (Object.keys(value).sort().join(',') !== 'type,x' || !Number.isInteger(value.x)) {
      throw new TypeError('invalid Point payload');
    }
    return { x: value.x, restored: true };
  }

  assert.deepEqual(
    JSON.parse('{"type":"Point","x":3}', restore),
    { x: 3, restored: true },
  );
  assert.throws(
    () => JSON.parse('{"type":"Point","x":"3","extra":true}', restore),
    /invalid Point payload/,
  );

  // reviver 不应根据输入字符串动态查找构造器或执行代码。
});

test('transfer 分离原 ArrayBuffer 及其既有视图', () => {
  const original = new ArrayBuffer(4);
  const originalView = new Uint8Array(original);
  originalView[0] = 9;

  const cloned = structuredClone(original, { transfer: [original] });

  assert.equal(original.byteLength, 0);
  assert.equal(originalView.byteLength, 0);
  assert.deepEqual([...new Uint8Array(cloned)], [9, 0, 0, 0]);

  // transfer 不是复制后继续共享；原 buffer 被 detach，既有 TypedArray 变为零长度。
});

File: concepts/06_text_binary_and_serialization/05_serialization_clone_and_transfer/python/test_02_numeric_and_trust_boundaries.py
"""数值模型、类型恢复与信任边界。

共同问题：通用数据格式是否保留语言数值类型；非标准数值如何处理；
自定义反序列化如何限制可构造类型和字段。
"""

# polyglot-family: text_binary_and_serialization
# polyglot-concept: serialization_clone_and_transfer
# polyglot-related: languages/python/stdlib/104-110_internet_data_handling/
# polyglot-related+: test_104_json_complete_encoding_decoding_streaming_and_cli_workflows.py

import json

import pytest


def test_json_keeps_python_integer_precision_but_changes_other_python_types():
    large = 9_007_199_254_740_993
    original = {1: (large,)}

    decoded = json.loads(json.dumps(original))

    assert decoded == {"1": [large]}
    assert decoded["1"][0] == large

    # Python int 可精确恢复该值，但 JSON number 没有统一精度上限；JavaScript Number
    # 消费者会丢精度。键和 tuple 也已分别变为字符串与 list。


def test_non_finite_numbers_require_an_explicit_interoperability_policy():
    assert json.dumps(float("nan")) == "NaN"

    with pytest.raises(ValueError, match="compliant"):
        json.dumps(float("nan"), allow_nan=False)

    # 默认 NaN 是 Python 扩展，不是标准 JSON；跨语言协议应使用 allow_nan=False。


def test_object_hook_restores_only_a_whitelisted_shape():
    class Point:
        def __init__(self, x):
            self.x = x

    def restore(value):
        if value.get("type") != "Point":
            return value
        if set(value) != {"type", "x"} or not isinstance(value["x"], int):
            raise ValueError("invalid Point payload")
        return Point(value["x"])

    decoded = json.loads('{"type":"Point","x":3}', object_hook=restore)
    assert isinstance(decoded, Point)
    assert decoded.x == 3

    with pytest.raises(ValueError, match="payload"):
        json.loads('{"type":"Point","x":"3","extra":true}', object_hook=restore)

    # object_hook 不应按输入名称动态 import/eval 类型；先验证标签、字段集合与值域。
